storage/aptosdb/src/fake_aptosdb.rs

```diff
@@ -0,0 +1,894 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    errors::AptosDbError,
+    gauged_api,
+    metrics::{LEDGER_VERSION, NEXT_BLOCK_EPOCH},
+    AptosDB,
+};
+use anyhow::{format_err, Result};
+use aptos_crypto::{hash::CryptoHash, HashValue};
+use aptos_storage_interface::{DbReader, DbWriter, ExecutedTrees, MAX_REQUEST_LIMIT};
+use aptos_types::{
+    access_path::AccessPath,
+    account_address::AccountAddress,
+    account_config::{AccountResource, NewBlockEvent},
+    contract_event::EventWithVersion,
+    epoch_state::EpochState,
+    event::{EventHandle, EventKey},
+    ledger_info::LedgerInfoWithSignatures,
+    proof::{
+        AccumulatorConsistencyProof, AccumulatorRangeProof, SparseMerkleProofExt,
+        TransactionAccumulatorProof, TransactionAccumulatorRangeProof,
+        TransactionAccumulatorSummary, TransactionInfoListWithProof, TransactionInfoWithProof,
+    },
+    state_proof::StateProof,
+    state_store::{
+        state_key::StateKey,
+        state_key_prefix::StateKeyPrefix,
+        state_storage_usage::StateStorageUsage,
+        state_value::{StateValue, StateValueChunkWithProof},
+        table,
+    },
+    transaction::{
+        Transaction, TransactionInfo, TransactionListWithProof, TransactionOutput,
+        TransactionOutputListWithProof, TransactionToCommit, TransactionWithProof, Version,
+    },
+    write_set::WriteSet,
+};
+use arc_swap::ArcSwapOption;
+use dashmap::DashMap;
+use itertools::zip_eq;
+use move_core_types::move_resource::MoveStructType;
+use std::sync::Arc;
+
+/// Provides a "fake" AptosDB.
+pub struct FakeAptosDB {
```

Comment:
It's hard to tell how much of this file is different to the regular DB? e.g., there appears to be a bunch of copy/pasted code, right? Is there any way to shrink this down, e.g., avoid the duplication?

Reply:
Ideally, we would want to "extend" AptosDB and just override the necessary methods, but I dont think it is possible in rust. I tried bunch of other things like implementing Deref, but it dint work because we are implementing multiple traits. I wanted to keep AptosDB untouched so this was an easy route. Let me know if there is anything else I should explore instead to make this better?

Reply:
Yeah, it's annoying that rust doesn't allow what you want to do... I'd check with @aptos-labs/storage. It might be better to just pull the common methods into a shared file.

Reply:
It can be a hell for maintenance like this but it's cleanest from the normal path point of view. 
To make things more right, we might need a pretty large refactor which extracts out interfaces for 1. storing factual data (via RocksDB vs Memory) and 2. authenticating the state (alternative to SMT/JMT ). But I don't think it's viable in short term or worth it. cc @zjma in case you think differently.

Based on that I'm gonna approve.

Reply:
It can be a hell for maintenance like this but it's cleanest from the normal path point of view.
To make things more right, we might need a pretty large refactor which extracts out interfaces for 1. storing factual data (via RocksDB vs Memory) and 2. authenticating the state (alternative to SMT/JMT ). But I don't think it's viable in short term or worth it. cc @zjma in case you think differently.

Based on that I'm gonna approve.